packages/block-editor/src/components/global-styles/border-panel.js

```diff
@@ -0,0 +1,285 @@
+} from '@wordpress/components';
+import { useCallback, useMemo } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import BorderRadiusControl from '../border-radius-control';
+import { useColorsPerOrigin } from './hooks';
+import { getValueFromVariable } from './utils';
+
+export function useHasBorderPanel( settings ) {
+	const controls = [
+		useHasBorderColorControl( settings ),
+		useHasBorderRadiusControl( settings ),
+		useHasBorderStyleControl( settings ),
+		useHasBorderWidthControl( settings ),
+	];
+
+	return controls.some( Boolean );
+}
+
+function useHasBorderColorControl( settings ) {
+	return settings?.border?.color;
+}
+
+function useHasBorderRadiusControl( settings ) {
+	return settings?.border?.radius;
+}
+
+function useHasBorderStyleControl( settings ) {
+	return settings?.border?.style;
+}
+
+function useHasBorderWidthControl( settings ) {
+	return settings?.border?.width;
+}
+
+function applyFallbackStyle( border ) {
+	if ( ! border ) {
+		return border;
+	}
+
+	if ( ! border.style && ( border.color || border.width ) ) {
+		return { ...border, style: 'solid' };
+	}
+
+	return border;
+}
+
+function applyAllFallbackStyles( border ) {
+	if ( ! border ) {
+		return border;
+	}
+
+	if ( hasSplitBorders( border ) ) {
+		return {
+			top: applyFallbackStyle( border.top ),
+			right: applyFallbackStyle( border.right ),
+			bottom: applyFallbackStyle( border.bottom ),
+			left: applyFallbackStyle( border.left ),
+		};
+	}
+
+	return applyFallbackStyle( border );
+}
+
+function BorderToolsPanel( {
+	resetAllFilter,
+	onChange,
+	value,
+	panelId,
+	children,
+} ) {
+	const resetAll = () => {
+		const updatedValue = resetAllFilter( value );
+		onChange( updatedValue );
+	};
+
+	return (
+		<ToolsPanel
+			label={ __( 'Border' ) }
+			resetAll={ resetAll }
+			panelId={ panelId }
+		>
+			{ children }
+		</ToolsPanel>
+	);
+}
+
+const DEFAULT_CONTROLS = {
+	radius: true,
+	color: true,
+	width: true,
+};
+
+export default function BorderPanel( {
+	as: Wrapper = BorderToolsPanel,
+	value,
+	onChange,
+	inheritedValue = value,
+	settings,
+	panelId,
+	defaultControls = DEFAULT_CONTROLS,
+} ) {
+	const colors = useColorsPerOrigin( settings );
+	const decodeValue = ( rawValue ) =>
+		getValueFromVariable( { settings }, '', rawValue );
+	const encodeColorValue = ( colorValue ) => {
+		const allColors = colors.flatMap(
+			( { colors: originColors } ) => originColors
+		);
+		const colorObject = allColors.find(
+			( { color } ) => color === colorValue
+		);
+		return colorObject
+			? 'var:preset|color|' + colorObject.slug
+			: colorValue;
+	};
+	const decodeColorValue = useCallback(
+		( colorValue ) => {
+			const allColors = colors.flatMap(
+				( { colors: originColors } ) => originColors
+			);
+			const colorObject = allColors.find(
+				( { slug } ) => colorValue === 'var:preset|color|' + slug
+			);
+			return colorObject ? colorObject.color : colorValue;
+		},
+		[ colors ]
+	);
```

Comment:
Note that I have to encode/decode colors every time in this component. Personally, I think that shouldn't happen in this component but instead it should be built-in in the lower level components `BorderBoxControl`... until `ColorPalette` because they all receive the "colors" presets as argument.

Reply:
Basically right now, if a user picks a "red" color from a palette, or picks a "red" color with the same value as a custom color, they are considered the same and automatically switched to "palette" colors. The problem is that the user intention might be different. 

 - If I pick a palette color, I expect the value to update as I update the palettes
 - If I pick a custom color, I expect the value to be static.

This behavior is broken now and the value is always dynamic but this is how it works in trunk too, so I kept it as is.